CAP/gap/WrapperCategory.gd

```diff
@@ -20,33 +22,142 @@
+#! More precisely, one has to take care of the following things:
+#!   * `W` should just be a reinterpretation of `cat_n` (with a nicer data structure), so we certainly
+#!     want "Wrapping" to be an equivalence of categories with pseudo-inverse "Unwrapping".
+#!   * `WrapperCategory` copies all properties from `cat_n` to `W`,
+#!     including properties like `IsSkeletalCategory` which are not necessarily preserved by mere equivalences.
+#!   * To fulfill the specification of WithGiven operations, wrapping a WithGiven object `A` in `cat_n`
+#!     as an object in `W` and unwrapping it again must give an object equal to `A`. So we require
+#!     "Wrapping" and then "Unwrapping" to be the identity. Conversely, let let `B_1` be an object in `W`
+#!     which we unwrap and wrap again to form an object `B_2`. By construction of `W`, `B_1` and `B_2` are equal
+#!     if and only if they are equal when unwrapped. But due to "Wrapping" and then "Unwrapping" being the identity,
+#!     unwrapping `B_2` simply gives an object equal to `B_1` unwrapped. Thus, also `B_1` and `B_2` are equal.
+#!     Thus, "Wrapping" has to be an equivalence which is a bijection on objects, and hence an isomorphism (although
+#!     "Unwrapping" is not necessarily its inverse).
+#!     Note: Alternatively, one can make sure that WithGiven operations in `cat_n` are only called
+#!     via the corresponding non-WithGiven operation in `cat_n`. This can be achieved by wrapping
+#!     all operations of `cat_n` (i.e. creating `W` with `only_primitive_operations := false`), disabling redirect functions
+#!     (i.e. creating `W` with the option `overhead := false`) and not calling WithGiven operations
+#!     of `W` manually.
+#!
+#! @Section How does it work?
+#!
+#! Operations in `WrapperCategory` are implemented as follows:
+#! <Enum>
+#! <Item>
+#!     Apply `object_datum` and `morphism_datum` to the input to get the underlying data.
+#! </Item>
+#! <Item>
+#!     Apply `modeling_tower_object_constructor` and `modeling_tower_morphism_constructor`
+#!     to the underlying data to get objects and morphisms in the tower.
+#! </Item>
+#! <Item>
+#!     Apply the operation of the tower.
+#! </Item>
+#! <Item>
+#!     Apply `modeling_tower_object_datum` or `modeling_tower_morphism_datum` to the result
+#!     to get the underlying data.
+#! </Item>
+#! <Item>
+#!     Apply `object_constructor` or `morphism_constructor` to the underlying data to get an
+#!     object or a morphism in the wrapper category.
+#! </Item>
+#! </Enum>
+#! The first two steps define the functor "Unwrapping" and the last two steps define the functor "Wrapping".
+#! "Wrapping" on objects and morphisms is called `ModeledObject` and `ModeledMorphism` in the code.
+#! "Unwrapping" on objects and morphisms is called `ModelingObject` and `ModelingMorphism` in the code.
+#!
+#! @Section Relation to `CompilerForCAP`
+#!
+#! The operation of the tower (step 3 above) usually unwraps objects and morphisms,
```

Comment:
This is confusion. In step 3 you are already working with the unwrapped objects/morphisms in cat_n.

Reply:
Ah, correct, what I mean is our usual kind of unwrapping which has nothing to do with WrapperCategory: obtaining the underlying data from a (tower) object or morphism. I have thought about calling this "(un)boxing" for some time because that seems to be the commonly used term for this concept: https://en.wikipedia.org/wiki/Boxing_(computer_science)

And while we're at it: I also don't like the term "WrapperCategory" anymore because "reinterpreting data"  is certainly not the same as "wrapping data". However, I currently have no good suggestion for a better name. Some obvious ideas:
* ReinterpretationCategory (or ReinterpretedCategory)
* TranslationCategory (or TranslatedCategory)
* RewritingCategory

In any case, we could keep the constructor "WrapperCategory" for the kind of wrapper category which really only wraps the data just as the original version of WrapperCategory did. This would maybe also disentangle some of the concepts which currently live in WrapperCategory.gi. For example, WrappingFunctor is only valid for this basic version of WrapperCategory and does not coincide with my definition of the functor "Wrapping" in the documentation.

Reply:
I agree and I like the name `ReinterpretationCategory` very much, this is the name I keep using when I explain the concept to others. What the documentation describes is indeed not a mere wrapping process, but a reinterpretation process. However, I still see `WrapperCategory` as a particular case of `ReinterpretationCategory` and the constructor's implementation confirms this; I would say the current implementation of the `WrappingFunctor` is broken.

Reply:
> I agree and I like the name `ReinterpretationCategory` very much

I also think "reinterpretation" is the term we want to use here, I'm just not convinced by `ReinterpretationCategory( C )` because spelling this out as "the reinterpretation category of C" sounds a bit weird. I would like to say `Reinterpretation( C )`, but the term `Reinterpretation` is too generic I think. `ReinterpretedCategory` has the problem that it could both mean the category being reinterpreted and the result of the reinterpretation. What about `ReinterpretationOfCategory` or `ReinterpretCategory`? I admit these sound less like category constructors, but then again, what else could they be? :D

Either way, we have to also fit this into the existing terms `ModelingObject`, `ModeledObject` etc. If we follow this scheme, we could rename `WrapperCategory` to `ModeledCategory`. But this would clash a bit with "model categories". Of course we could also rename `ModeledObject` to `ReinterpretedObject`, and maybe keep `ModelingObject`.

The functor from `cat_n` to `W` would best be called "Reinterpretation", I think. How would we call its inverse: "OriginalInterpretation", "Model" (or "ReReinterpretation" :D )?

Disclaimer: I'm very in favor of renaming WrapperCategory, I just want to make sure we keep things consistent, that's why I'm bringing up so many questions :D

Reply:
I'm sorry for the non-constructive nature of the following comment: Please do not opt for the name "translation category". There is a concept of a translation groupoid associated to a group acting on a set, and a natural extension of this idea would be to call translation category a category associated to a monoid acting on a set. I would like to reserve this name for that particular concept.

Reply:
What about `ReinterpretedCategory`?

Reply:
> What about `ReinterpretedCategory`?

No, `ReinterpretedCategory` could also be a name for the current `ModelingCategory`. Hmm.

Reply:
> No, `ReinterpretedCategory` could also be a name for the current `ModelingCategory`. Hmm.

Exactly, that's what I meant above:
> `ReinterpretedCategory` has the problem that it could both mean the category being reinterpreted [i.e. the current `ModelingCategory`] and the result of the reinterpretation.

Reply:
We just discussed the following naming convention:

* `WrapperCategory`, `ModeledObject` and `ModeledMorphism` are renamed to `ReinterpretationOf{Category,Object,Morphism}`.
* `Modeling{Category,Object,Morphism}` are kept.
* `WrappingFunctor` is renamed to `ReinterpretationFunctor` with component functions `ReinterpretationOf{Object,Morphism}`.

So only one question remains: How to call the inverse of `ReinterpretationFunctor`? Ideas:
* `ModelingFunctor` or `ModelFunctor`
* `OriginalInterpretationFunctor`, `InitialInterpretationFunctor` or `PrimaryInterpretationFunctor`
* `ReverseInterpretationFunctor`
* `InverseInterpretationFunctor` or `InterpretationFunctorInverse`

Reply:
The problem I see with `ModelingFunctor` is that if one interprets it in the same way as `Modeling{Category,Object,Morphism}`, then one would expect `ModelingFunctor( F )` to return a functor which somehow models `F`. In this sense, it would not be consistent with the other names. `ModelFunctor` what be perfect, I think, but that clashes with "model categories" again :/

Reply:
Oo Sorry, I thought the `ReinterpretationFunctor` and _ModelingFunctor_ are operations that take as input the reinterpretation category and output equivalence functors between the the modeling category and its reinterpretation category. But as you mentioned, the input is a functor which then goes a completely different direction.

Reply:
No, you are right: `ReinterpretationFunctor` and `ModelingFunctor` get the reinterpretation of a category as input, I confused myself :D Still, when writing `ModelingFunctor( C )` one might expect to get a functor modeling `C` somehow.

But I just noticed: There is no `UnwrappingFunctor` in the code right now, so I only need the term for the documentation. And in the documentation I think it's not confusing to call the functor "Model(ing)" because the context is clear. So everything should be fine :-)

Reply:
> So only one question remains: How to call the inverse of `ReinterpretationFunctor`? Ideas:
> 
> * `ModelingFunctor` or `ModelFunctor`
> * `OriginalInterpretationFunctor`, `InitialInterpretationFunctor` or `PrimaryInterpretationFunctor`
> * `ReverseInterpretationFunctor`
> * `InverseInterpretationFunctor` or `InterpretationFunctorInverse`

I like `InverseInterpretationFunctor`.

Reply:
Maybe my impressions help (as I'm pretty much an outsider with this).

I dislike these:

- `OriginalInterpretationFunctor`, `InitialInterpretationFunctor` or `PrimaryInterpretationFunctor`

Only judging from their names, I see their connection with the concept, but I would probably have to rethink their meaning every time I would use them; they seem easy to forget and a bit unintuitive.

On the other hand, I like these names:
- `ReverseInterpretationFunctor`
- `InverseInterpretationFunctor` or `InterpretationFunctorInverse`

they have an `Inverse` in their names and that tells me right away I'm dealing with some sort of equivalence -> this is easy to remember and to understand. But of course, it depends on how much you like this emphasis on being an inverse.

But I'm an amateur, so feel free to ignore this. :D

Reply:
On the contrary, all your comments are to the point.